File: wox.plugin.python/src/wox_plugin/models/query.py
"""
Wox Query Models

This module provides models for handling user queries in Wox plugins.

Queries represent user input and the context in which they are made,
including the query text, selection (text/files), environment information,
and trigger keywords.
"""

import json
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Optional


class SelectionType(str, Enum):
    """
    Enumeration of selection types for user-selected content.

    When a user makes a selection in another application and invokes Wox,
    the selection can be either text or files.
    """

    TEXT = "text"
    """
    Text selection.

    Represents selected text content from another application.
    The actual text is available in the Selection.text field.
    """

    FILE = "file"
    """
    File selection.

    Represents one or more selected file paths from the file system.
    The file paths are available in the Selection.file_paths field.
    """


class QueryType(str, Enum):
    """
    Enumeration of query types in Wox.

    Defines how the query was initiated and what content it contains.
    """

    INPUT = "input"
    """
    Input query triggered by typing in the Wox search box.

    This is the most common query type, triggered when the user
    activates Wox and types text or uses a trigger keyword.
    """

    SELECTION = "selection"
    """
    Selection query triggered by selecting content and invoking Wox.

    This query type occurs when the user selects text or files in
    another application and then invokes Wox (e.g., via hotkey).
    The selection data is available in Query.selection.
    """


@dataclass
class MetadataCommand:
    """
    Command metadata for registering plugin commands.

    Used with `api.register_query_commands()` to register commands
    that can be invoked by typing specific keywords or patterns.

    Attributes:
        command: The command keyword/trigger
        description: Human-readable description of what the command does

    Example usage:
        # Register commands in init()
        commands = [
            MetadataCommand(
                command="todo",
                description="Manage your todo list"
            ),
            MetadataCommand(
                command="note",
                description="Quick note taking"
            ),
        ]
        await api.register_query_commands(ctx, commands)
    """

    command: str
    """
    The command keyword/trigger string.

    Users can invoke this command by typing this keyword in Wox.
    Example: "todo", "note", "calc"
    """

    description: str
    """
    Human-readable description of the command.

    This is shown to users to help them understand what the
    command does. Should be concise and informative.
    """

    def to_json(self) -> str:
        """
        Convert to JSON string with camelCase naming.

        Returns:
            JSON string representation
        """
        return json.dumps(
            {
                "Command": self.command,
                "Description": self.description,
            }
        )

    @classmethod
    def from_json(cls, json_str: str) -> "MetadataCommand":
        """
        Create from JSON string with camelCase naming.

        Args:
            json_str: JSON string with "Command" and "Description"

        Returns:
            A new MetadataCommand instance
        """
        data = json.loads(json_str)
        return cls(
            command=data.get("Command", ""),
            description=data.get("Description", ""),
        )


@dataclass
class Selection:
    """
    User-selected content from another application.

    Represents text or files that the user selected before invoking Wox.
    This allows plugins to operate on content from other applications.

    Attributes:
        type: The type of selection (TEXT or FILE)
        text: The selected text content (when type is TEXT)
        file_paths: List of selected file paths (when type is FILE)

    Example usage:
        async def query(ctx: Context, query: Query) -> List[Result]:
            if query.type == QueryType.SELECTION:
                if query.selection.type == SelectionType.TEXT:
                    # Process selected text
                    text = query.selection.text
                    return [process_text(text)]
                elif query.selection.type == SelectionType.FILE:
                    # Process selected files
                    files = query.selection.file_paths
                    return [process_files(files)]
    """

    type: SelectionType = field(default=SelectionType.TEXT)
    """
    The type of selection.

    Either TEXT (selected text content) or FILE (selected file paths).
    """

    text: str = field(default="")
    """
    The selected text content.

    Contains the actual text that the user selected in another application.
    Only populated when type is TEXT.
    """

    file_paths: List[str] = field(default_factory=list)
    """
    List of selected file paths.

    Contains the full paths to files that the user selected.
    Only populated when type is FILE.
    """

    def to_json(self) -> str:
        """
        Convert to JSON string with camelCase naming.

        Returns:
            JSON string representation
        """
        return json.dumps(
            {
                "Type": self.type,
                "Text": self.text,
                "FilePaths": self.file_paths,
            }
        )

    @classmethod
    def from_json(cls, json_str: str) -> "Selection":
        """
        Create from JSON string with camelCase naming.

        Args:
            json_str: JSON string containing selection data

        Returns:
            A new Selection instance
        """
        data = json.loads(json_str)

        if not data.get("Type"):
            data["Type"] = SelectionType.TEXT

        return cls(
            type=SelectionType(data.get("Type")),
            text=data.get("Text", ""),
            file_paths=data.get("FilePaths", []),
        )

    def __str__(self) -> str:
        """
        Convert selection to string representation.

        Returns:
            Text content for TEXT type, comma-joined paths for FILE type,
            or empty string if no data
        """
        if self.type == SelectionType.TEXT and self.text:
            return self.text
        elif self.type == SelectionType.FILE and self.file_paths:
            return ",".join(self.file_paths)
        return ""


@dataclass
class QueryEnv:
    """
    Environment information at the time of the query.

    Provides context about the user's current environment when the
    query was made, including the active window and browser URL.

    Attributes:
        active_window_title: Title of the currently active window
        active_window_pid: Process ID of the active window (0 if unavailable)
        active_window_icon: Icon of the active window (as WoxImage dict)
        active_browser_url: URL from the active browser tab

    Example usage:
        # Get context about the current application
        window_title = query.env.active_window_title
        if "Chrome" in window_title:
            # User is browsing, show browser-related results
    """

    active_window_title: str = field(default="")
    """
    Title of the active window when the query was made.

    Contains the window title of the application that was active
    when the user invoked Wox. Can be used to provide context-aware results.

    Examples: "Visual Studio Code", "Google Chrome", "Untitled - Notepad"
    """

    active_window_pid: int = field(default=0)
    """
    Process ID of the active window.

    The PID of the application that was active when the query was made.
    Can be used to interact with the active application programmatically.

    Note: May be 0 if the PID cannot be determined.
    """

    active_window_icon: dict = field(default_factory=dict)
    """
    Icon of the active window as a WoxImage dictionary.

    Contains the icon of the active application window. The format
    is a WoxImage serialized to dictionary format.

    Note: May be empty if no icon is available.
    """

    active_browser_url: str = field(default="")
    """
    URL from the active browser tab.

    Contains the URL of the currently open tab if:
    1. The active window is a supported browser (Chrome, Edge, etc.)
    2. The Wox Chrome Extension is installed

    The extension is available at:
    https://github.com/Wox-launcher/Wox.Chrome.Extension

    Note: Only available when both conditions above are met.
    """

    def to_json(self) -> str:
        """
        Convert to JSON string with camelCase naming.

        Returns:
            JSON string representation
        """
        return json.dumps(
            {
                "ActiveWindowTitle": self.active_window_title,
                "ActiveWindowPid": self.active_window_pid,
                "ActiveWindowIcon": self.active_window_icon,
                "ActiveBrowserUrl": self.active_browser_url,
            }
        )

    @classmethod
    def from_json(cls, json_str: str) -> "QueryEnv":
        """
        Create from JSON string with camelCase naming.

        Args:
            json_str: JSON string containing environment data

        Returns:
            A new QueryEnv instance
        """
        data = json.loads(json_str)
        return cls(
            active_window_title=data.get("ActiveWindowTitle", ""),
            active_window_pid=data.get("ActiveWindowPid", 0),
            active_window_icon=data.get("ActiveWindowIcon", {}),
            active_browser_url=data.get("ActiveBrowserUrl", ""),
        )


@dataclass
class Query:
    """
    A user query in Wox.

    Represents a complete query from the user, including the query text,
    type, selection, environment context, and parsed components like
    trigger keyword and search terms.

    Attributes:
        id: Unique identifier for this query instance
        type: The type of query (INPUT or SELECTION)
        raw_query: The full raw query text as typed by the user
        selection: Selected content (for SELECTION type queries)
        env: Environment context information
        trigger_keyword: The trigger keyword if present (empty for global queries)
        command: The command keyword after trigger keyword
        search: The search text after command (empty if only command)

    Example usage:
        async def query(ctx: Context, query: Query) -> List[Result]:
            # For a query like ">todo add buy groceries"
            # trigger_keyword = "todo"
            # command = "add"
            # search = "buy groceries"

            if query.trigger_keyword == "todo":
                return handle_todo_command(query.command, query.search)
            elif query.is_global_query():
                return handle_global_search(query.search)
    """

    id: str
    """
    Unique identifier for this query instance.

    Can be used to track and correlate requests, especially useful
    for logging and debugging.
    """

    type: QueryType
    """
    The type of query.

    INPUT: User typed in the Wox search box
    SELECTION: User selected content and invoked Wox
    """

    raw_query: str
    """
    The full raw query text as typed by the user.

    Contains the complete query string including trigger keyword,
    command, and search text. For example: ">todo add buy groceries"
    """

    selection: Selection
    """
    Selected content for SELECTION type queries.

    Contains the text or file paths that the user selected before
    invoking Wox. Empty for INPUT type queries.
    """

    env: QueryEnv
    """
    Environment context at query time.

    Contains information about the active window, browser URL,
    and other contextual information.
    """

    trigger_keyword: str = field(default="")
    """
    The trigger keyword for plugin-specific queries.

    When a plugin registers with a trigger keyword (e.g., ">todo"),
    this field contains that keyword. Empty for global queries.

    Example: For ">todo add item", trigger_keyword = "todo"
    """

    command: str = field(default="")
    """
    The command keyword after the trigger keyword.

    For queries with a command structure, this contains the first
    word after the trigger keyword.

    Example: For ">todo add item", command = "add"
    """

    search: str = field(default="")
    """
    The search text after the command.

    Contains the remaining text after extracting the trigger keyword
    and command. Used as the actual search/query parameter.

    Example: For ">todo add buy groceries", search = "buy groceries"
    """

    def to_json(self) -> str:
        """
        Convert to JSON string with camelCase naming.

        Returns:
            JSON string representation
        """
        return json.dumps(
            {
                "QueryId": self.id,
                "Type": self.type,
                "RawQuery": self.raw_query,
                "Selection": json.loads(self.selection.to_json()),
                "Env": json.loads(self.env.to_json()),
                "TriggerKeyword": self.trigger_keyword,
                "Command": self.command,
                "Search": self.search,
            }
        )

    @classmethod
    def from_json(cls, json_str: str) -> "Query":
        """
        Create from JSON string with camelCase naming.

        Args:
            json_str: JSON string containing query data

        Returns:
            A new Query instance
        """
        data = json.loads(json_str)

        if not data.get("Type"):
            data["Type"] = QueryType.INPUT

        return cls(
            id=data.get("QueryId", ""),
            type=QueryType(data.get("Type")),
            raw_query=data.get("RawQuery", ""),
            selection=Selection.from_json(data.get("Selection", Selection().to_json())),
            env=QueryEnv.from_json(data.get("Env", QueryEnv().to_json())),
            trigger_keyword=data.get("TriggerKeyword", ""),
            command=data.get("Command", ""),
            search=data.get("Search", ""),
        )

    def is_global_query(self) -> bool:
        """
        Check if this is a global query without trigger keyword.

        Global queries are queries that don't have a trigger keyword,
        meaning they're handled by the global search system rather
        than a specific plugin.

        Returns:
            True if this is a global INPUT query without trigger keyword

        Example:
            if query.is_global_query():
                # Handle global search
                return search_everywhere(query.search)
        """
        return self.type == QueryType.INPUT and not self.trigger_keyword

    def __str__(self) -> str:
        """
        Convert query to string representation.

        Returns:
            The raw query text for INPUT queries, the selection string
            for SELECTION queries, or empty string
        """
        if self.type == QueryType.INPUT:
            return self.raw_query
        elif self.type == QueryType.SELECTION:
            return str(self.selection)
        return ""


@dataclass
class ChangeQueryParam:
    """
    Parameters for changing the current query in Wox.

    Used with `api.change_query()` to programmatically change the
    query text and type.

    Attributes:
        query_type: The type of query to change to (INPUT or SELECTION)
        query_text: The new query text (for INPUT type)
        query_selection: The new selection (for SELECTION type)

    Example usage:
        # Change to a new input query
        await api.change_query(ctx, ChangeQueryParam(
            query_type=QueryType.INPUT,
            query_text="new search text"
        ))

        # Change to a selection query
        await api.change_query(ctx, ChangeQueryParam(
            query_type=QueryType.SELECTION,
            query_selection=Selection(
                type=SelectionType.TEXT,
                text="selected text"
            )
        ))
    """

    query_type: QueryType
    """
    The type of query to change to.

    Either INPUT (for text-based queries) or SELECTION (for selection-based).
    """

    query_text: str = field(default="")
    """
    The new query text for INPUT type queries.

    The text that will appear in the Wox search box.
    """

    query_selection: Selection = field(default_factory=Selection)
    """
    The new selection for SELECTION type queries.

    Contains the text or file selection to set.
    """

    def to_json(self) -> str:
        """
        Convert to JSON string with camelCase naming.

        Returns:
            JSON string representation
        """
        data = {
            "QueryType": self.query_type,
            "QueryText": self.query_text,
        }
        if self.query_selection:
            data["QuerySelection"] = json.loads(self.query_selection.to_json())
        return json.dumps(data)

    @classmethod
    def from_json(cls, json_str: str) -> "ChangeQueryParam":
        """
        Create from JSON string with camelCase naming.

        Args:
            json_str: JSON string containing change query parameters

        Returns:
            A new ChangeQueryParam instance
        """
        data = json.loads(json_str)

        if not data.get("QueryType"):
            data["QueryType"] = QueryType.INPUT

        return cls(
            query_type=QueryType(data.get("QueryType")),
            query_text=data.get("QueryText", ""),
            query_selection=Selection.from_json(data.get("QuerySelection", Selection().to_json())),
        )


@dataclass
class RefreshQueryParam:
    """
    Parameters for refreshing the current query.

    Used with `api.refresh_query()` to re-execute the current query
    with the same text, optionally preserving the selected result index.

    This is useful when plugin data changes and you want to update the
    displayed results without the user having to retype the query.

    Attributes:
        preserve_selected_index: Whether to keep the current selection position

    Example usage:
        # After deleting an item, refresh and reset selection
        await api.refresh_query(ctx, RefreshQueryParam(
            preserve_selected_index=False  # Reset to first item
        ))

        # After updating an item, refresh and keep selection
        await api.refresh_query(ctx, RefreshQueryParam(
            preserve_selected_index=True  # Keep current position
        ))
    """

    preserve_selected_index: bool = field(default=False)
    """
    Controls whether to maintain the previously selected item index after refresh.

    When True, the user's current selection index in the results list is preserved.
    This is useful when updating results without disrupting the user's position.

    When False, the selection resets to the first item (index 0).
    Use this when the selected item may have been removed or repositioned.

    Examples:
        - preserve_selected_index=True: After marking an item as favorite,
          the results list updates but the user stays on the same item
        - preserve_selected_index=False: After deleting the selected item,
          the selection moves to the first item
    """


class CopyType(str, Enum):
    """
    Enumeration of clipboard copy types.

    Defines the type of content to copy to the system clipboard.
    """

    TEXT = "text"
    """
    Copy text content.

    The text will be copied as plain text to the clipboard.
    """

    IMAGE = "image"
    """
    Copy image content.

    The image (as WoxImage) will be copied to the clipboard as image data.
    """


@dataclass
class CopyParams:
    """
    Parameters for copying content to the system clipboard.

    Used with `api.copy()` to copy text or images to the clipboard.

    Attributes:
        type: The type of content to copy (TEXT or IMAGE)
        text: The text content to copy (for TEXT type)
        wox_image: The WoxImage dict to copy (for IMAGE type)

    Example usage:
        # Copy text to clipboard
        await api.copy(ctx, CopyParams(
            type=CopyType.TEXT,
            text="Hello, World!"
        ))

        # Copy image to clipboard
        await api.copy(ctx, CopyParams(
            type=CopyType.IMAGE,
            wox_image=WoxImage.new_absolute("/path/to/image.png").to_dict()
        ))
    """

    type: CopyType = field(default=CopyType.TEXT)
    """
    The type of content to copy.

    Either TEXT (plain text) or IMAGE (WoxImage).
    """

    text: str = field(default="")
    """
    The text content to copy.

    Contains the text that will be copied to the clipboard.
    Only used when type is TEXT.
    """

    wox_image: Optional[dict] = field(default=None)
    """
    The WoxImage dictionary to copy.

    Contains a WoxImage serialized to dictionary format.
    Only used when type is IMAGE.
    """

    def to_json(self) -> str:
        """
        Convert to JSON string with camelCase naming.

        Returns:
            JSON string representation
        """
        return json.dumps(
            {
                "type": self.type,
                "text": self.text,
                "woxImage": self.wox_image,
            }
        )

File: wox.core/plugin/query.go
package plugin

import (
	"context"
	"strings"
	"wox/common"
	"wox/setting/definition"
	"wox/util/selection"

	"github.com/samber/lo"
)

type QueryResultActionType = string
type QueryType = string
type QueryVariable = string
type QueryResultTailType = string
type QueryResultTailTextCategory = string
type QueryRefinementType = string

const (
	QueryTypeInput     QueryType = "input"     // user input query
	QueryTypeSelection QueryType = "selection" // user selection query
)

const (
	QueryResultActionTypeExecute QueryResultActionType = "execute"
	QueryResultActionTypeForm    QueryResultActionType = "form"
)

const (
	QueryVariableSelectedText     QueryVariable = "{wox:selected_text}"
	QueryVariableSelectedFile     QueryVariable = "{wox:selected_file}"
	QueryVariableActiveBrowserUrl QueryVariable = "{wox:active_browser_url}"
	QueryVariableFileExplorerPath QueryVariable = "{wox:file_explorer_path}"
)

const (
	QueryResultTailTypeText  QueryResultTailType = "text"  // string type
	QueryResultTailTypeImage QueryResultTailType = "image" // WoxImage type
)

const (
	QueryResultTailTextCategoryDefault QueryResultTailTextCategory = "default"
	QueryResultTailTextCategoryDanger  QueryResultTailTextCategory = "danger"
	QueryResultTailTextCategoryWarning QueryResultTailTextCategory = "warning"
	QueryResultTailTextCategorySuccess QueryResultTailTextCategory = "success"
)

const (
	QueryRefinementTypeSingleSelect QueryRefinementType = "singleSelect"
	QueryRefinementTypeMultiSelect  QueryRefinementType = "multiSelect"
	QueryRefinementTypeToggle       QueryRefinementType = "toggle"
	QueryRefinementTypeSort         QueryRefinementType = "sort"
)

// Query from Wox. See "Doc/Query.md" for details.
type Query struct {
	// Id identifies the current query session from UI.
	// It can be used to correlate async updates with the active query.
	Id string

	// SessionId identifies the UI session that owns this query.
	// It stays stable for the lifetime of a UI instance.
	SessionId string

	// By default, Wox will only pass QueryTypeInput query to plugin.
	// plugin author need to enable MetadataFeatureQuerySelection feature to handle QueryTypeSelection query
	Type QueryType

	// Raw query, this includes trigger keyword if it has.
	// We didn't recommend use this property directly. You should always use Search property.
	RawQuery string

	// Trigger keyword of a query. It can be empty if user is using global trigger keyword.
	// Empty trigger keyword means this query will be a global query, see IsGlobalQuery.
	//
	// NOTE: Only available when query type is QueryTypeInput
	TriggerKeyword string

	// Command part of a query.
	// Empty command means this query doesn't have a command.
	//
	// NOTE: Only available when query type is QueryTypeInput
	Command string

	// Search part of a query.
	// Empty search means this query doesn't have a search part.
	Search string

	// User selected or drag-drop data, can be text or file or image etc
	//
	// NOTE: Only available when query type is QueryTypeSelection
	Selection selection.Selection

	// additional query environment data
	// expose more context env data to plugin, E.g. plugin A only show result when active window title is "Chrome"
	Env QueryEnv

	// Refinements carries query-scoped UI state selected by the user. Values are
	// strings to keep the plugin-facing API close to Env-style key/value data;
	// multi-select refinements are encoded as comma-separated strings.
	Refinements map[string]string

	// ContextData carries hidden query-scoped data. Unlike Refinements, this is
	// not rendered by the UI and is intended for plugin handoffs such as a shell
	// working directory.
	ContextData common.ContextData
}

func (q *Query) IsGlobalQuery() bool {
	return q.Type == QueryTypeInput && q.TriggerKeyword == ""
}

func (q *Query) String() string {
	if q.Type == QueryTypeInput {
		return q.RawQuery
	}
	if q.Type == QueryTypeSelection {
		return q.Selection.String()
	}
	return ""
}

type QueryEnv struct {
	ActiveWindowTitle string          // active window title when user query, empty if not available
	ActiveWindowPid   int             // active window pid when user query, 0 if not available
	ActiveWindowIcon  common.WoxImage // active window icon when user query, empty if not available

	// active browser url when user query
	// Only available when active window is browser and https://github.com/Wox-launcher/Wox.Chrome.Extension is installed
	ActiveBrowserUrl string

	// These fields are core-only for built-in system plugins. Do not add them
	// to SDK models or public plugin docs unless we explicitly decide to expose
	// the native window handle/dialog state as public plugin API.
	ActiveWindowId               string `json:"-"` // exact top-level window id; Windows HWND, macOS CGWindowID
	ActiveWindowIsOpenSaveDialog bool   `json:"-"` // active window is open/save dialog when user query
}

// QueryResponse is the complete plugin answer for one query execution.
// The old API returned only []QueryResult, which forced query-scoped UI data
// such as refinements and layout hints into side channels. Keeping them in one
// response lets the UI apply the controls and results from the same query id.
type QueryResponse struct {
	Results     []QueryResult
	Refinements []QueryRefinement
	Layout      QueryLayout
	Context     QueryContext
}

func NewQueryResponse(results []QueryResult) QueryResponse {
	return QueryResponse{Results: results}
}

// QueryLayout carries query-scoped presentation hints.
// These fields are pointers because zero is a meaningful value for some hints:
// ResultPreviewWidthRatio=0 intentionally gives the preview the full result
// area. The old metadata side request could return that value explicitly; the
// QueryResponse path must keep the same distinction between unset and zero.
type QueryLayout struct {
	Icon                    *common.WoxImage                 `json:"Icon,omitempty"`
	ResultPreviewWidthRatio *float64                         `json:"ResultPreviewWidthRatio,omitempty"`
	GridLayout              *MetadataFeatureParamsGridLayout `json:"GridLayout,omitempty"`
}

// QueryContext carries the backend's canonical classification for a query.
// The UI can make a quick local guess for immediate rendering, but only core
// has the final parser state after shortcuts and trigger-keyword matching.
type QueryContext struct {
	IsGlobalQuery bool   `json:"IsGlobalQuery"`
	PluginId      string `json:"PluginId"`
}

// BuildQueryContext centralizes the parser result that the UI cannot reproduce
// cheaply. In particular, shortcuts and trigger-keyword parsing happen in core,
// so the response carries the resolved global/plugin classification back.
func BuildQueryContext(query Query, queryPlugin *Instance) QueryContext {
	queryContext := QueryContext{IsGlobalQuery: query.IsGlobalQuery()}
	if !queryContext.IsGlobalQuery && queryPlugin != nil {
		queryContext.PluginId = queryPlugin.Metadata.Id
	}
	return queryContext
}

// QueryRefinement describes one query-scoped control such as type filters or
// sort modes. Options are deliberately simple values because plugins, not core,
// interpret the selected values when the next query is executed.
type QueryRefinement struct {
	Id           string
	Title        string
	Type         QueryRefinementType
	Options      []QueryRefinementOption
	DefaultValue []string
	Hotkey       string
	Persist      bool
}

type QueryRefinementOption struct {
	Value    string
	Title    string
	Icon     common.WoxImage
	Keywords []string
	Count    *int
}

// RefreshQueryParam contains parameters for refreshing a query
type RefreshQueryParam struct {
	// PreserveSelectedIndex controls whether to maintain the previously selected item index after refresh
	// When true, the user's current selection index in the results list is preserved
	// When false, the selection resets to the first item (index 0)
	PreserveSelectedIndex bool
}

const QueryResultDragDataTypeFiles = "files"

// QueryResultDragData declares data the UI can export through a native drag session.
type QueryResultDragData struct {
	Type  string
	Files []string
}

// Query result return from plugin
type QueryResult struct {
	// Result id, should be unique. It's optional, if you don't set it, Wox will assign a random id for you
	Id string
	// Title support i18n
	Title string
	// SubTitle support i18n
	SubTitle string
	Icon     common.WoxImage
	Preview  WoxPreview
	// Score of the result, the higher the score, the more relevant the result is, more likely to be displayed on top
	Score int64
	// ScoreKey is an optional stable identity for actioned-result scoring when title or subtitle is dynamic.
	ScoreKey string
	// Group results, Wox will group results by group name
	Group string
	// Score of the group, the higher the score, the more relevant the group is, more likely to be displayed on top
	GroupScore int64
	// Tails are additional results associate with this result, can be displayed in result detail view
	Tails   []QueryResultTail
	Actions []QueryResultAction
	// DragData declares what can be dragged out of Wox for this result.
	DragData *QueryResultDragData
}

type QueryResultTail struct {
	// Tail id, should be unique. It's optional, if you don't set it, Wox will assign a random id for you
	Id           string
	Type         QueryResultTailType
	Text         string                      // only available when type is QueryResultTailTypeText
	TextCategory QueryResultTailTextCategory // optional semantic category when type is QueryResultTailTypeText
	Image        common.WoxImage             // only available when type is QueryResultTailTypeImage
	ImageWidth   *float64                    // optional width for image tails
	ImageHeight  *float64                    // optional height for image tails
	// Tooltip is hover-only context for compact tails. This keeps visual-only tails
	// readable without forcing plugins to trade result width for explanatory text.
	Tooltip string
	// Additional data associate with this tail, can be retrieved later
	ContextData map[string]string

	// internal use
	IsSystemTail bool
}

func NewQueryResultTailText(text string) QueryResultTail {
	return NewQueryResultTailTextWithCategory(text, QueryResultTailTextCategoryDefault)
}

func NewQueryResultTailTextWithCategory(text string, category QueryResultTailTextCategory) QueryResultTail {
	if category == "" {
		category = QueryResultTailTextCategoryDefault
	}

	return QueryResultTail{
		Type:         QueryResultTailTypeText,
		Text:         text,
		TextCategory: category,
	}
}

func NewQueryResultTailTexts(texts ...string) []QueryResultTail {
	return lo.Map(texts, func(text string, index int) QueryResultTail {
		return NewQueryResultTailText(text)
	})
}

type QueryResultAction struct {
	// Action id, should be unique. It's optional, if you don't set it, Wox will assign a random id for you
	Id string
	// Action type, use QueryResultActionTypeExecute for immediate execution and QueryResultActionTypeForm for form submission
	Type QueryResultActionType
	// Name support i18n
	Name string
	Icon common.WoxImage
	// If true, Wox will use this action as default action. There can be only one default action in results
	// This can be omitted, if you don't set it, Wox will use the first action as default action
	IsDefault bool
	// If true, Wox will not hide after user select this result
	PreventHideAfterAction bool
	// Hotkey to trigger this action. E.g. "ctrl+Shift+Space", "Ctrl+1", "Command+K"
	// Case insensitive, space insensitive
	// If IsDefault is true, Hotkey will be set to enter key by default
	// Wox will normalize the hotkey to platform specific format. E.g. "ctrl" will be converted to "control" on macOS
	Hotkey string
	// Additional data associate with this action, can be retrieved later
	ContextData map[string]string

	// For execute action
	Action func(ctx context.Context, actionContext ActionContext) `json:"-"` // Exclude from JSON serialization

	// For form action
	Form     definition.PluginSettingDefinitions
	OnSubmit func(ctx context.Context, actionContext FormActionContext) `json:"-"` // Exclude from JSON serialization

	// internal use
	IsSystemAction bool
}

type ActionContext struct {
	// The ID of the result that triggered this action
	// This is automatically set by Wox when the action is invoked
	// Useful for calling UpdateResult API to update the result's UI
	ResultId string

	// The ID of the action that was triggered
	// This is automatically set by Wox when the action is invoked
	// Useful for calling UpdateResul API to update this specific action's UI
	ResultActionId string

	// Additional data associate with this action
	ContextData common.ContextData
}

type FormActionContext struct {
	ActionContext
	Values map[string]string
}

func (q *QueryResult) ToUI() QueryResultUI {
	return QueryResultUI{
		Id:         q.Id,
		Title:      q.Title,
		SubTitle:   q.SubTitle,
		Icon:       q.Icon,
		Preview:    q.Preview,
		Score:      q.Score,
		Group:      q.Group,
		GroupScore: q.GroupScore,
		Tails:      q.Tails,
		DragData:   q.DragData,
		Actions: lo.Map(q.Actions, func(action QueryResultAction, index int) QueryResultActionUI {
			actionType := action.Type
			if actionType == "" {
				actionType = QueryResultActionTypeExecute
			}
			return QueryResultActionUI{
				Id:                     action.Id,
				Type:                   actionType,
				Name:                   action.Name,
				Icon:                   action.Icon,
				IsDefault:              action.IsDefault,
				PreventHideAfterAction: action.PreventHideAfterAction,
				Hotkey:                 action.Hotkey,
				Form:                   action.Form,
				ContextData:            action.ContextData,
				IsSystemAction:         action.IsSystemAction,
			}
		}),
	}
}

func (q *QueryResponse) ToUI() QueryResponseUI {
	return QueryResponseUI{
		Results: lo.Map(q.Results, func(result QueryResult, index int) QueryResultUI {
			return result.ToUI()
		}),
		Refinements: q.Refinements,
		Layout:      q.Layout,
		Context:     q.Context,
	}
}

type QueryResultUI struct {
	QueryId    string
	Id         string
	Title      string
	SubTitle   string
	Icon       common.WoxImage
	Preview    WoxPreview
	Score      int64
	Group      string
	GroupScore int64
	Tails      []QueryResultTail
	Actions    []QueryResultActionUI
	DragData   *QueryResultDragData
	IsGroup    bool
}

type QueryResponseUI struct {
	Results             []QueryResultUI
	Refinements         []QueryRefinement
	Layout              QueryLayout
	Context             QueryContext
	QueryStartTimestamp int64 // end-to-end query start timestamp, preferably from Flutter request send time
}

// PushResultsPayload is used to push additional results to UI for a query.
type PushResultsPayload struct {
	QueryId string
	Results []QueryResultUI
}

type QueryResultActionUI struct {
	Id                     string
	Type                   QueryResultActionType
	Name                   string
	Icon                   common.WoxImage
	IsDefault              bool
	PreventHideAfterAction bool
	Hotkey                 string
	Form                   definition.PluginSettingDefinitions
	ContextData            map[string]string

	// internal use
	IsSystemAction bool
}

// UpdatableResult is used to update a query result that is currently displayed in the UI.
//
// This struct serves two purposes:
// 1. As the return type of GetUpdatableResult() - contains the current state of the result
// 2. As the parameter of UpdateResult() - specifies which fields to update
//
// When returned by GetUpdatableResult():
// - All fields contain the current values from the result cache
// - You can modify any fields and pass it back to UpdateResult()
//
// When passed to UpdateResult():
// - All fields except Id are optional (pointers). Only non-nil fields will be updated.
// - This allows you to update specific fields without affecting others.
//
// Example usage:
//
//	// Get current result state
//	updatableResult := api.GetUpdatableResult(ctx, resultId)
//	if updatableResult != nil {
//	    // Modify any fields
//	    newTitle := "Downloading... 50%"
//	    updatableResult.Title = &newTitle
//	    updatableResult.Tails = append(*updatableResult.Tails, NewQueryResultTailText("50%"))
//
//	    // Update the result
//	    api.UpdateResult(ctx, *updatableResult)
//	}
type UpdatableResult struct {
	// Id is required - identifies which result to update
	Id string

	// Optional fields - only non-nil fields will be updated
	Title    *string
	SubTitle *string
	Icon     *common.WoxImage
	Preview  *WoxPreview
	Tails    *[]QueryResultTail
	Actions  *[]QueryResultAction
	DragData *QueryResultDragData
}

// store latest result value after query/refresh, so we can retrieve data later in action/refresh
type QueryResultCache struct {
	Result         QueryResult // store the full QueryResult including actions with callbacks
	PluginInstance *Instance
	Query          Query
	Layout         QueryLayout // query layout used when polishing this result, so later updates keep the same surface sizing
	// FlushBatch is the debouncer batch that first sent this result in a visible response.
	FlushBatch int
	// BatchQueueElapsed is the elapsed time when queryRun put this result into the debouncer queue.
	BatchQueueElapsed    int64
	BatchQueueElapsedSet bool
	// QueryElapsed is the elapsed time when queryRun received the plugin response, measured from the end-to-end query start.
	QueryElapsed    int64
	QueryElapsedSet bool
	// PluginQueryElapsed is only the raw Plugin.Query duration, excluding manager polish and UI conversion.
	PluginQueryElapsed    int64
	PluginQueryElapsedSet bool
}

func newQueryInputWithPlugins(query string, pluginInstances []*Instance) (Query, *Instance) {
	var terms = strings.Split(query, " ")
	if len(terms) == 0 {
		return Query{
			Type:     QueryTypeInput,
			RawQuery: query,
		}, nil
	}

	var rawQuery = query
	var triggerKeyword, command, search string
	var possibleTriggerKeyword = terms[0]
	var mustContainSpace = strings.Contains(query, " ")

	pluginInstance, found := lo.Find(pluginInstances, func(instance *Instance) bool {
		return lo.Contains(instance.GetTriggerKeywords(), possibleTriggerKeyword)
	})
	if found && (mustContainSpace) {
		// non global trigger keyword
		triggerKeyword = possibleTriggerKeyword

		if len(terms) == 1 {
			// no command and search
			command = ""
			search = ""
		} else {
			if len(terms) == 2 {
				// e.g "wpm install", we treat "install" as search, only "wpm install " will be treated as command
				command = ""
				search = terms[1]
			} else {
				var possibleCommand = terms[1]
				if lo.ContainsBy(pluginInstance.GetQueryCommands(), func(item MetadataCommand) bool {
					return item.Command == possibleCommand
				}) {
					// command and search
					command = possibleCommand
					search = strings.Join(terms[2:], " ")
				} else {
					// no command, only search
					command = ""
					search = strings.Join(terms[1:], " ")
				}
			}
		}
	} else {
		// non trigger keyword
		triggerKeyword = ""
		command = ""
		search = rawQuery
		pluginInstance = nil
	}

	return Query{
		Type:           QueryTypeInput,
		RawQuery:       query,
		TriggerKeyword: triggerKeyword,
		Command:        command,
		Search:         search,
	}, pluginInstance
}
